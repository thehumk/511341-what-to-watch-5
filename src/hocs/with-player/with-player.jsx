import {connect} from 'react-redux';
import {propsForFilms, propsForRouter} from '../../utils/prop-types';

export const withPlayer = (Component) => {
  class WithPlayer extends React.PureComponent {
    constructor(props) {
      super(props);

      this.films = props.films;
      this.routerProps = props.routerProps;
      this.film = this.films.find((elem) => elem.id.toString() === this.routerProps.match.params.id);

      this.state = {
        playFilm: true,
        progressVideo: 0,
        timeLeft: 0,
      };

      this.videoRef = React.createRef();

      this.exitButtonClickHandler = this.exitButtonClickHandler.bind(this);
      this.filmPlayHandler = this.filmPlayHandler.bind(this);
      this.filmPauseHandler = this.filmPauseHandler.bind(this);
      this.fullScreenClickHandler = this.fullScreenClickHandler.bind(this);
      this.timeUpdateHandler = this.timeUpdateHandler.bind(this);
    }

    componentDidMount() {
      const video = this.videoRef.current;
      video.play();
    }

    componentDidUpdate() {
      const video = this.videoRef.current;
      this.state.playFilm ? video.play() : video.pause();
    }

    exitButtonClickHandler() {
      this.routerProps.history.goBack();
    }

    filmPlayHandler() {
      this.setState({playFilm: true});
    }

    filmPauseHandler() {
      this.setState({playFilm: false});
    }

    fullScreenClickHandler() {
      const video = this.videoRef.current;
      video.requestFullscreen();
    }

    timeUpdateHandler() {
      const video = this.videoRef.current;

      this.setState({
        progressVideo: video.currentTime * 100 / video.duration,
        timeLeft: video.duration - video.currentTime,
      });
    }

    render() {
      return (
        <Component
          {...this.props}
          film={this.film}
          playFilm={this.state.playFilm}
          timeLeftFilm={this.state.timeLeft}
          progressVideo={this.state.progressVideo}
          exitButtonClickHandler={this.exitButtonClickHandler}
          filmPlayHandler={this.filmPlayHandler}
          filmPauseHandler={this.filmPauseHandler}
          fullScreenClickHandler={this.fullScreenClickHandler}>
          <video
            ref={this.videoRef}
            src={this.film.video_link}
            className="player__video"
            poster={`img/${this.film.bigPoster}`}
            onTimeUpdate={this.timeUpdateHandler}
            muted>
          </video>
        </Component>
      );
    }
  }

  WithPlayer.propTypes = {
    films: PropTypes.arrayOf(propsForFilms).isRequired,
    routerProps: propsForRouter,
  };

  const mapStateToProps = ({DATA}) => ({
    films: DATA.films,
  });

  return connect(mapStateToProps)(WithPlayer);
};
