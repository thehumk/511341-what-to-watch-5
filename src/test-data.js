export const noop = () => {};

export const MATCH = {params: {id: `1`}};

export const mockComments = [
  {
    comment: `123`,
    date: `2020-10-29T13:38:44.769Z`,
    id: 1,
    rating: 4.6,
    user: {
      id: 12,
      name: `Isaac`
    },
  },
  {
    comment: `1234`,
    date: `2020-10-29T13:38:45.769Z`,
    id: 2,
    rating: 4.5,
    user: {
      id: 13,
      name: `Isa`
    },
  }
];

/* eslint-disable */

export const mockFilms = [
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Action`,
    id: 1,
    is_favorite: true,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Drama`,
    id: 2,
    is_favorite: false,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
];

/* eslint-enable */

export const state = {
  DATA: {
    films: mockFilms,
    promoFilm: mockFilms[0],
    genresList: [`All`, `Action`, `Drama`],
    isApplicationReady: false,
  },
  FILMS_STATUS: {
    activeGenre: `All`,
    renderFilmsCount: 8,
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    user: null,
  },
  COMMENTS: {
    comments: mockComments,
  },
};
