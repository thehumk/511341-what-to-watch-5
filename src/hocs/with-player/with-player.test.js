import React from 'react';
import PropTypes from 'prop-types';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import {withPlayer} from './with-player';
import {noop, mockFilms, state, MATCH} from '../../test-data';

const mockStore = configureMockStore()(state);

const MockComponent = (props) => {
  return <div>{props.children}</div>;
};

MockComponent.propTypes = {
  children: PropTypes.node,
};

const MockComponentContainer = withPlayer(MockComponent);

it(`Should WithPlayer render correctly`, () => {
  const tree = renderer.create((
    <Provider store={mockStore}>
      <MockComponentContainer
        films={mockFilms}
        match={MATCH}
        onExitClick={noop}>
        <React.Fragment/>
      </MockComponentContainer>
    </Provider>
  ), {
    createNodeMock() {
      return {
        play: noop,
        pause: noop,
      };
    }
  }).toJSON();

  expect(tree).toMatchSnapshot();
});
