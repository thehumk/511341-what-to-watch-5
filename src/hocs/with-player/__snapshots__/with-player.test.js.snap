// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should WithPlayer render correctly 1`] = `
<div>
  <video
    className="player__video"
    muted={true}
    onTimeUpdate={[Function]}
    poster="img/undefined"
    src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
  />
</div>
`;
